Appsembler/ficus/feature/registration default field overrides

bryanlandia opened the pull request:
The purpose of this is to make the registration and login forms more extensible/configurable via `SiteConfiguration`.  It allows (when using combined login/registration):

- custom prologue and epilogue text for login and registration forms
  - pass `LOGIN_FORM_PROLOGUE`, `LOGIN_FORM_EPILOGUE`,  or `REGISTRATION_FORM_PROLOGUE`, `REGISTRATION_FORM_EPILOGUE` as entries in `SiteConfiguration` object.  
- custom form field order for login and registration forms.  Change order fields display in the forms.
  - pass `LOGIN_FORM_FIELD_ORDER` or `REGISTRATION_FORM_FIELD_ORDER` as entries in `SiteConfiguration` object.   
- override of registration form field properties (using facility already provided in edx-platform for external auth)
  -  This can be used to modify existing fields on the registration page; i.e., to change help text or placeholder value.  It must be a path to a class, in dot-separated syntax; e.g., `custom_form_app.forms.RegistrationFieldOverrides`.  It can be defined in `settings. REGISTRATION_FIELD_OVERRIDES`, so is platform wide at this point.
- Allow `hidden` input field types in registration and login forms, so that unwanted default fields that can't readily be removed from the form can at least be hidden from the user.  They can then be made optional.

It has been used for the CMC and main InterSystems Ficus sites in staging, so far.

tkeemon commented on openedx/core/djangoapps/user_api/helpers.py:
```diff
@@ -132,27 +134,34 @@ class FormDescription(object):
         forms.Textarea: "textarea",
         forms.BooleanField: "checkbox",
-        forms.EmailField: "email",
+        forms.EmailField: "email"
```
Is there any particular reason why you're removing the trailing commas in this file?

bryanlandia replied:
@tkeemon No, I didn't mean to commit any of the changes to this file, actually.  Thanks for catching it.

bryanlandia replied:
Nope, this was unintentional.

tkeemon commented on openedx/core/djangoapps/user_api/helpers.py:
```diff
@@ -13,6 +13,8 @@
 from django.utils.encoding import force_text
 from django.utils.functional import Promise
 
+from openedx.core.djangoapps.site_configuration import helpers as configuration_helpers
```
I don't think this is used anywhere in this file.

thraxil replied:
Pardon me butting in here. What's the general consensus on static analysis (pep8, flake8, etc)? It looks like circle CI is running pep8 on this repo and finding a bunch of issues, but they're being ignored. Does upstream really not care about it? Do we care about it?

I ask partly because the two comments here are both things that flake8 would've identified automatically.

natea commented:
@thraxil see here: https://edx.readthedocs.io/projects/edx-developer-guide/en/latest/style_guides/python-guidelines.html

More general info about contributing to Open edX:
https://edx.readthedocs.io/projects/edx-developer-guide/en/latest/process/index.html

this is a good overview of how new features are reviewed and merged upstream: https://openedx.atlassian.net/wiki/spaces/COMM/pages/44073040/2015+Open+edX+Presentations?preview=/44073040/46793633/Contributing%20to%20Open%20edX.pptx

thraxil commented:
@natea, Yeah, I've seen the style guidelines and contributor documentation.  I guess my question is more about the builds failing here. If you click through to the failing CircleCI build, you'll see (among others):

    Build failed running pavelib.quality.run_pep8: Too many pep8 violations. Number of pep8 violations: 78

`./scripts/all-tests.sh` appears to come from upstream and most of the code with pep8 violations is from upstream. Where are they all coming from? Does edX just not actually enforce their style guidelines? Why have a test script that runs pep8 and pylint if it's just ignored?

I don't want to block this PR or sidetrack too much. I'm just trying to get a sense of what our view of the situation is. I mean, as far as I can tell, all of the [CircleCI builds for edx-platform](https://circleci.com/gh/appsembler/edx-platform) fail consistently because of this.  So I'm wondering what the point of running them is? Do *we* consider it a problem or is it something that we're fine just ignoring?

abeals commented:
@johnbaldwin you may want to take a look at this PR, since there's a decent overlap with what you're trying to do with reports + custom profile fields.

bryanlandia commented:
thanks for the comments @thraxil and @tkeemon .  As @johnbaldwin has moved forward with a Python style guide, I won't pick up that conversation here, but I've fixed the issues pointed out in this conversation.

@tkeemon I rebased against appsembler/ficus/develop, fixed the style and import issue, and squashed that fix commit.  Can you take another look?

bryanlandia commented:
@tkeemon I'd like to get this merged for the Trinity Ficus upgrade.  Could you take another look?

bryanlandia commented:
@tkeemon Thanks!  You're right, I'll work on some tests and kick off some review from edX to see if I can get it upstreamed.

bryanlandia commented:
@tkeemon Following up here.  I think at this point, since we ended up not using it for CMC Ficus, I'll probably target an upstream PR for Hawthorne and backport to Ginkgo.

OmarIthawi commented:
@bryanlandia I'll close this PR. Please feel free to repoen if it is still relevant.
